### PR TITLE
eventbus add tls config when connect to mqtt

### PR DESCRIPTION
**Pull request by lvchenggang**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
There is a possibility that eventbus connect to mqtt broker need security verify. 
This patch provide a optional function that if we config EventBusTls.Enable=true, we could through a security way connect to mqtt broker.
If we config EventBusTls.Enable=false, then business as usual.

**Special notes for your reviewer**:
This patch do not attention the certificates generate of mqtt broker and client, because these certificates are generally generated by mqtt broker.

**Does this PR introduce a user-facing change?**:
NONE

### CODE REVIEW

File: edge/pkg/eventbus/common/util/common.go [hunk 1]
```diff
@@ -50,8 +53,39 @@ func HubClientInit(server, clientID, username, password string) *MQTT.ClientOpti
 			opts.SetPassword(password)
 		}
 	}
-	tlsConfig := &tls.Config{InsecureSkipVerify: true, ClientAuth: tls.NoClientCert}
+
+	klog.Infof("Start to config MQTT client tls connect")
+	tlsConfig := &tls.Config{}
+	if eventconfig.Config.Tls.Enable == true {
```
Comment (fisherxu) [inline edge/pkg/eventbus/common/util/common.go:59]:
```suggestion
	if eventconfig.Config.Tls.Enable {
```

File: pkg/apis/componentconfig/edgecore/v1alpha1/default.go [hunk 2]
```diff
@@ -214,6 +214,12 @@ func NewMinEdgeCoreConfig() *EdgeCoreConfig {
 				MqttServerExternal: "tcp://127.0.0.1:1883",
 				MqttServerInternal: "tcp://127.0.0.1:1884",
 				MqttMode:           MqttModeExternal,
+				Tls: &EventBusTls{
```
Comment (fisherxu) [inline pkg/apis/componentconfig/edgecore/v1alpha1/default.go:217]: move it to default config, no need set in minConfig

File: pkg/apis/componentconfig/edgecore/v1alpha1/types.go [hunk 3]
```diff
@@ -344,6 +344,23 @@ type EventBus struct {
 	// +Required
 	// default: 2
 	MqttMode MqttMode `json:"mqttMode"`
+	// Tls indicates tls config for EventBus module
+	Tls *EventBusTls `json:"tls,omitempty"`
+}
+
+type EventBusTls struct {
+	// Enable indicates whether enable this protocol
```
Comment (fisherxu) [inline pkg/apis/componentconfig/edgecore/v1alpha1/types.go:352]:
```suggestion
	// Enable indicates whether enable tls connection
```

File: pkg/apis/componentconfig/edgecore/v1alpha1/types.go [hunk 4]
```diff
@@ -344,6 +344,23 @@ type EventBus struct {
...
+	// default false
+	Enable bool `json:"enable,omitempty"`
+	// TLSMqttCAFile set ca file path
```
Comment (fisherxu) [inline pkg/apis/componentconfig/edgecore/v1alpha1/types.go:355]:
```suggestion
	// TLSMqttCAFile sets ca file path
```

File: edge/pkg/eventbus/common/util/common.go [hunk 5]
```diff
@@ -50,8 +53,39 @@ func HubClientInit(server, clientID, username, password string) *MQTT.ClientOpti
 			opts.SetPassword(password)
 		}
 	}
-	tlsConfig := &tls.Config{InsecureSkipVerify: true, ClientAuth: tls.NoClientCert}
+
+	klog.Infof("Start to config MQTT client tls connect")
```
Comment (fisherxu) [inline edge/pkg/eventbus/common/util/common.go:57]:
```suggestion
	klog.V(4).Infof("Start to set TLS configuration for MQTT client")
```

File: pkg/apis/componentconfig/edgecore/v1alpha1/types.go [hunk 6]
```diff
@@ -344,6 +344,23 @@ type EventBus struct {
 	// +Required
 	// default: 2
 	MqttMode MqttMode `json:"mqttMode"`
+	// Tls indicates tls config for EventBus module
+	Tls *EventBusTls `json:"tls,omitempty"`
```
Comment (fisherxu) [inline pkg/apis/componentconfig/edgecore/v1alpha1/types.go:348]:
```suggestion
	TLS *EventBusTLS `json:"eventBusTLS ,omitempty"`
```

File: pkg/apis/componentconfig/edgecore/v1alpha1/types.go [hunk 7]
```diff
@@ -344,6 +344,23 @@ type EventBus struct {
 	// +Required
 	// default: 2
 	MqttMode MqttMode `json:"mqttMode"`
+	// Tls indicates tls config for EventBus module
```
Comment (fisherxu) [inline pkg/apis/componentconfig/edgecore/v1alpha1/types.go:347]:
```suggestion
	// TLS indicates tls config for EventBus module
```

File: edge/pkg/eventbus/common/util/common.go [hunk 8]
```diff
@@ -50,8 +53,39 @@ func HubClientInit(server, clientID, username, password string) *MQTT.ClientOpti
...
+		tlsConfig = &tls.Config{
+			RootCAs:            pool,
+			Certificates:       []tls.Certificate{cert},
+			InsecureSkipVerify: false,
+		}
+	} else {
+		tlsConfig = &tls.Config{InsecureSkipVerify: true, ClientAuth: tls.NoClientCert}
+	}
 	opts.SetTLSConfig(tlsConfig)
+	klog.Infof("config MQTT client tls connect successful")
```
Comment (fisherxu) [inline edge/pkg/eventbus/common/util/common.go:87]:
```suggestion
	klog.Infof("set TLS configuration for MQTT client successfully")
```

**Review by fisherxu — COMMENTED** [review]
Thanks for the enhancement :) just some nits

File: edge/pkg/eventbus/common/util/common.go [hunk 9]
```diff
@@ -50,8 +53,39 @@ func HubClientInit(server, clientID, username, password string) *MQTT.ClientOpti
...
+		tlsConfig = &tls.Config{
+			RootCAs:            pool,
+			Certificates:       []tls.Certificate{cert},
+			InsecureSkipVerify: false,
+		}
+	} else {
+		tlsConfig = &tls.Config{InsecureSkipVerify: true, ClientAuth: tls.NoClientCert}
+	}
 	opts.SetTLSConfig(tlsConfig)
+	klog.Infof("set TLS configuration for MQTT client successfully")
```
Comment (fisherxu) [inline edge/pkg/eventbus/common/util/common.go:87]: klog.V(4).

File: pkg/apis/componentconfig/edgecore/v1alpha1/types.go [hunk 10]
```diff
@@ -344,6 +344,23 @@ type EventBus struct {
 	// +Required
 	// default: 2
 	MqttMode MqttMode `json:"mqttMode"`
+	// TLS indicates tls config for EventBus module
+	TLS *EventBusTLS `json:"eventBusTLS ,omitempty"`
```
Comment (fisherxu) [inline pkg/apis/componentconfig/edgecore/v1alpha1/types.go:348]:
```suggestion
	TLS *EventBusTLS `json:"eventBusTLS,omitempty"`
```

File: pkg/apis/componentconfig/edgecore/v1alpha1/types.go [hunk 11]
```diff
@@ -344,6 +344,23 @@ type EventBus struct {
...
+}
+
+type EventBusTLS struct {
```
Comment (fisherxu) [inline pkg/apis/componentconfig/edgecore/v1alpha1/types.go:351]: Need comment for the struct

File: edge/pkg/eventbus/common/util/common.go (hunk 1 above)
Comment (lvchenggang) [reply edge/pkg/eventbus/common/util/common.go]: thanks for your comments, now these problems has been modified

**Review by fisherxu — COMMENTED** [review]
Overall lgtm, btw, could you please:
1. squash the commits 
2. add your email addr in the commit to your github personal email setting, or github can't count your contributions.
image

File: edge/pkg/eventbus/common/util/common.go (hunk 9 above)
Comment (lvchenggang) [reply edge/pkg/eventbus/common/util/common.go]: thanks for your remind, now the commits has been squashed and my email addr has been added

**Comment by fisherxu** [comment]
Add the email addr here :)
image

**Comment by lvchenggang** [comment]
> Add the email addr here :)
> image

image
the primary email has already been added

**Comment by fisherxu** [comment]
> the primary email has already been added

But github still can't recognise your commit, looks like email addr in commit is not your primary email addr

**Comment by fisherxu** [comment]
Need to fix the UT of eventbus

**Comment by lvchenggang** [comment]
> Need to fix the UT of eventbus

Thanks for point out, the UT of eventbus has been modified.

File: edge/pkg/eventbus/common/util/common.go [hunk 12]
```diff
@@ -2,11 +2,14 @@ package util
...
 	"crypto/tls"
+	"crypto/x509"
 	"errors"
+	"io/ioutil"
 	"os"
 	"time"
 
 	MQTT "github.com/eclipse/paho.mqtt.golang"
+	eventconfig "github.com/kubeedge/kubeedge/edge/pkg/eventbus/config"
 	"k8s.io/klog"
```
Comment (fisherxu) [inline edge/pkg/eventbus/common/util/common.go:13]:
```suggestion
import (
	"crypto/tls"
	"crypto/x509"
	"errors"
	"io/ioutil"
	"os"
	"time"

	MQTT "github.com/eclipse/paho.mqtt.golang"
	"k8s.io/klog"
	
	eventconfig "github.com/kubeedge/kubeedge/edge/pkg/eventbus/config"
```

File: edge/pkg/eventbus/common/util/common_test.go [hunk 13]
```diff
@@ -19,6 +19,8 @@ package util
 import (
 	"crypto/tls"
 	"fmt"
+	eventconfig "github.com/kubeedge/kubeedge/edge/pkg/eventbus/config"
+	"github.com/kubeedge/kubeedge/pkg/apis/componentconfig/edgecore/v1alpha1"
 	"net/url"
 	"reflect"
 	"testing"
```
Comment (fisherxu) [inline edge/pkg/eventbus/common/util/common_test.go:26]: Sort the imports.

Comment (lvchenggang) [reply edge/pkg/eventbus/common/util/common_test.go:26]: ok

**Pull request merged**